src/freenet/node/DNSRequester.java

```diff
@@ -61,60 +90,316 @@ public void run() {
     }
 
     private void realRun() {
-        // run DNS requests for not recently checked, unconnected
-        // nodes to avoid the coupon collector's problem
-        PeerNode[] nodesToCheck = Arrays.stream(node.getPeers().myPeers())
-            .filter(peerNode -> !peerNode.isConnected())
+        long now = System.currentTimeMillis();
+
+        // Get all peers and partition them
+        PeerNode[] allPeers = node.getPeers().myPeers();
+        PartitionedPeers partitioned = partitionPeers(allPeers);
+        cleanupStalePeers(now, partitioned);
+
+        if (logMINOR) {
+            if ((now - lastLogTime) > LOG_THROTTLE_MS) {
+                Logger.minor(this, "Processing Requests: " + (partitioned.hostnamePeers.size() + partitioned.ipPeers.size()) + " unconnected peers.");
+                lastLogTime = now;
+            }
+        }
+
+        boolean isStartupPhase = (now < node.getStartupTime() + ATTEMPT_WINDOW_MS);
+
+        // ===== PHASE A: IP peers (safe to be fast) =====
+        processIpOnlyPeers(now, partitioned.ipPeers, isStartupPhase);
+
+        // ===== PHASE B: Hostname peers (stealth path to protect anonymity) =====
+        boolean touchedHostname = processHostnamePeers(now, partitioned.hostnamePeers, isStartupPhase);
+
+        // If the startup phase just ended, clear the startup-only set to release memory.
+        if (!isStartupPhase && !startupPhaseCompleted) {
+            startupPhaseCompleted = true; // Mark as completed so we only do this once.
+            if(logMINOR) {
+                Logger.minor(this, "Startup phase complete. Clearing " + peersSeenDuringStartup.size() + " startup peer references.");
+            }
+            peersSeenDuringStartup.clear();
+        }
+
+        // ===== SLEEP TIME =====
+        int nextMaxWaitTime = determineNextWaitTime(touchedHostname);
+
+        try {
+            synchronized (this) {
+                wait(nextMaxWaitTime);
+            }
+        } catch (InterruptedException e) {
+            // Ignore, just wake up. Just sleeping to not busy wait anyway
+        }
+    }
+
+    /**
+     * The state-tracking maps are cleaned up when a peer connects. However, if a peer is permanently
+     * removed from the network (i.e., it no longer appears in node.getPeers().myPeers()), its entry
+     * might persist in these maps indefinitely, leading to a slow memory leak. This helps keep maps clean.
+     * @param now The current time in milliseconds.
+     * @param partitioned The list of all unconnected peers (both IP and hostname).
+     */
+    private void cleanupStalePeers(long now, PartitionedPeers partitioned) {
+        if (now - lastCleanupTime > CLEANUP_INTERVAL_MS) {
```

Comment:
Is this limiting necessary? And could this get the PeerNodes directly from `node.getPeers().myPeers()` so it would not have to collect them into one list again?